src/core/memory.d

```diff
@@ -847,6 +847,104 @@ struct GC
     {
         gc_runFinalizers( segment );
     }
+
+    /**
+     * Queries the GC whether the current thread is running object finalization
+     * as part of a GC collection.
+     *
+     * As some GC implementations (such as the current conservative one) don't
+     * support GC memory allocation during object finalization, this function
+     * can be used to guard against such programming errors.
+     *
+     * Returns:
+     *  true if the current thread is in a finalizer, a destructor invoked by
+     *  the GC.
+     */
+    static bool inFinalizer() nothrow @nogc @safe
+    {
+        return gc_inFinalizer();
+    }
+
+    ///
+    @safe nothrow @nogc unittest
+    {
+        // Although unit tests can trigger garbage collection,
+        // and subsequent object finalization, it's highly unlikely
+        // that top-level `unittest` statements are executed from
+        // a GC finalization.
+        assert(!GC.inFinalizer);
+    }
+
+    ///
+    unittest
+    {
+        enum Outcome
+        {
+            notExecuted,
+            executedManually,
+            executedByTheGC
+        }
+
+        static class Resource
+        {
+            static Outcome outcome;
+
+            this()
+            {
+                outcome = Outcome.notExecuted;
+            }
+
+            ~this()
+            {
+                if (GC.inFinalizer)
+                {
+                    outcome = Outcome.executedByTheGC;
+
+                    import core.exception : InvalidMemoryOperationError;
+                    try
+                    {
+                        /*
+                         * Presently, allocating GC memory during finalization
+                         * is forbidden and leads to
+                         * `InvalidMemoryOperationError` being thrown.
+                         *
+                         * `GC.inFinalizer` can be used to guard against
+                         * programming erros such as these and is also a more
+                         * efficient way to verify whether a destructor was
+                         * invoked by the GC.
+                         */
+                        cast(void) GC.malloc(1);
+                        assert(false);
+                    }
+                    catch (InvalidMemoryOperationError e)
+                    {
+                        return;
+                    }
+                    assert(false);
+                }
+                else
+                    outcome = Outcome.executedManually;
+            }
+        }
+
+        static void createGarbage()
+        {
+            auto r = new Resource;
+            r = null;
+        }
+
+        assert(Resource.outcome == Outcome.notExecuted);
+        createGarbage();
+        GC.collect;
+        assert(
+            Resource.outcome == Outcome.notExecuted ||
+            Resource.outcome == Outcome.executedByTheGC);
```

Comment:
This sucks, but I don't know how to guarantee something becomes garbage.

Reply:
Not completely the same, but you could call runFinalizers explicitly with a range that contains the address of the dtor. Also used here:
https://github.com/dlang/druntime/blob/2429fb822ff11ec5692deea6805abb32f75cbc7f/src/rt/lifetime.d#L2634-L2636

Reply:
Actually, initially, before reading the docs on `runFinalizers` I was planning on using it, as I thought that it works on data, but after reading that it works on code segments, I wasn't sure anymore that it fits the bill. 

Given that that I need to update the docs to mention `runFinalizers`, I'll also add it to the unit test.